### PR TITLE
ddl: Support drop column on composite index

### LINKED ISSUE #19196: ddl: support dropping columns with composite indices
## Feature Request

**Is your feature request related to a problem? Please describe:**
After PR #18852, #18812 is supported in TiDB.

Now we'd like to go on improving the `ALTER TABLE ... DROP COLUMN` to support dropping columns with composite indices like:

```
tidb>  create table t(c1 int, c2 int, c3 int, index idx_c1_c2_c3(c1, c2, c3));
Query OK, 0 rows affected (0.02 sec)

tidb> alter table t drop column idx_c1_c2_c3;
ERROR 8200 (HY000): can't drop column a with index covered now
```

**Describe the feature you'd like:**
See #3364 

**Describe alternatives you've considered:**
Manually, the `ALTER TABLE ... DROP COLUMN` in TiDB can be archived by 2 separate steps:
1. Drop all indices that cover the column to be dropped.
2. Drop the column without any index.

However, for some third-party frameworks that relies on dropping columns directly, workarounds have to be done for this limitation. This can be disappointing.

**Teachability, Documentation, Adoption, Migration Strategy:**

More help is available by referring:

* [The internal algorithm of TiDB online DDL](https://github.com/pingcap/tidb/blob/master/docs/design/2018-10-08-online-DDL.md), more specifically [AddColumn](https://github.com/pingcap/tidb/blob/master/docs/design/2018-10-08-online-DDL.md#add-column-operation) action and [AddIndex](https://github.com/pingcap/tidb/blob/master/docs/design/2018-10-08-online-DDL.md#add-index-operation) action.

* The key schema state change codes:
  * [onDropColumns()](https://github.com/pingcap/tidb/blob/56fd348d20cbc1983fa5afa88a62192cc01d59fa/ddl/column.go#L383)
  * [onDropIndex()](https://github.com/pingcap/tidb/blob/56fd348d20cbc1983fa5afa88a62192cc01d59fa/ddl/index.go#L586)
  * [applyDiff()](https://github.com/pingcap/tidb/blob/56fd348d20cbc1983fa5afa88a62192cc01d59fa/infoschema/builder.go#L41), which updates the information schema served as a schema cache in TiDB server.

### PR DESCRIPTION
**Pull request by blacktear23**
### What problem does this PR solve?

Issue Number: close #19196

Problem Summary:

For now TiDB can drop column with single column index covered. This PR will let TiDB can drop column with composite index covered. But if the column is covered by cluster index this PR still not support this case.

### What is changed and how it works?

On drop column with composite index, the step is like below:

1. Create new composite index without dropped column with temp name, and not set the index's state to `StatePublic` after write reorganization.
2. Drop column and column covered single column indexes.
3. At column's `StateDeleteReorganization` state rename temp composite index to origin name and set those indexes to `StatePublic`
4. Finish the job.

For tracking the new composite indexes reorg progress, there has 2 fields add in `job.Model`:

* NeedReorgIndexIDs string: The composite index's IDs in string separated by '|' (For example: "12|13|15")
* CurrentReorgIndexIdx int: The index for NeedReorgIndexIDs array

After introduced those 2 fields, we can reuse the create index reorg code. Just check if all indexes is reorg finished then fall-through to do the rest works. And if we reuse the create index reorg logic we can get the benefit of `TxnMerge` and `Ingest`  support.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Support drop column with composite index covered.
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -243,9 +247,15 @@ testpkg: failpoint-enable
 ifeq ("$(pkg)", "")
 	@echo "Require pkg parameter"
 else
+ifeq ("$(case)", "")
 	@echo "Running unit test for github.com/pingcap/tidb/$(pkg)"
 	@export log_level=fatal; export TZ='Asia/Shanghai'; \
-	$(GOTEST) -ldflags '$(TEST_LDFLAGS)' -cover github.com/pingcap/tidb/$(pkg) || { $(FAILPOINT_DISABLE); exit 1; }
+	$(GOTEST) -tags 'intest' -v -ldflags '$(TEST_LDFLAGS)' -cover github.com/pingcap/tidb/$(pkg) -test.run "TestDropColumnWithUniqCompositeIndex" || { $(FAILPOINT_DISABLE); exit 1; }
```
Comment (bb7133) [inline Makefile:253]: Please remove this debug code.

Comment (blacktear23) [reply Makefile:253]: We must add the `-tags 'intest'` other wise we cannot run any test code. And I think this should be a bug on Makefile.

File: ddl/column_with_index.go [hunk 2]
```diff
@@ -0,0 +1,683 @@
...
+	_, cidxInfos := listIndexesWithColumn(colName.L, tblInfo.Indices)
+	if len(cidxInfos) == 0 {
+		return false, nil
+	}
+	colMaps := getColumnMapFromTable(tblInfo)
+
+	for _, idxInfo := range cidxInfos {
+		for _, icol := range idxInfo.Columns {
+			if col, have := colMaps[icol.Name.L]; have && mysql.HasAutoIncrementFlag(col.GetFlag()) {
+				if col.Name.L == colName.L && ctx.GetSessionVars().AllowRemoveAutoInc {
```
Comment (blacktear23) [inline ddl/column_with_index.go:586]: There has a corner case: If TiDB allow drop auto increment column and the composite index only has this column is auto increment shall we allow this column to be dropped?

For example:
```
create table t (a int auto_increment, b int, c int, index i_ab(a, b));
alter table t drop column a;
```
shall we allow drop the column `a`?

File: ddl/column_with_index.go [hunk 3]
```diff
@@ -0,0 +1,683 @@
...
+			SQLMode:       job.ReorgMeta.SQLMode,
+			Warnings:      job.ReorgMeta.Warnings,
+			WarningsCount: job.ReorgMeta.WarningsCount,
+			Location:      job.ReorgMeta.Location,
+			ReorgTp:       job.ReorgMeta.ReorgTp,
+			IsDistReorg:   job.ReorgMeta.IsDistReorg,
+		}
+		// Reset reorg meta and snapshot version
+		job.ReorgMeta = newReorgMeta
+		job.SnapshotVer = 0
```
Comment (blacktear23) [inline ddl/column_with_index.go:558]: I have a question here. Set SnapshotVer to zero will re-initialize the `reorgInfo`. At `reorgInfo` initialize function, the snapshot version will get from store. So I don't know the SnapshotVer is same as prev index reorg process. And if the SnapshotVer is not same is there has some issue at reorg process? Such as missing some column to create the index?

File: Makefile [hunk 4]
```diff
@@ -243,9 +247,15 @@ testpkg: failpoint-enable
 ifeq ("$(pkg)", "")
 	@echo "Require pkg parameter"
 else
+ifeq ("$(case)", "")
```
Comment (Defined2014) [inline Makefile:250]: Please don't change Makefile in feature related PR. I think it's better to open a new PR for it.

File: ddl/column_with_index.go (hunk 2 above)
Comment (Defined2014) [reply ddl/column_with_index.go:584]: I think we should allow drop column `a`. What is your concerned about this case?

File: Makefile (hunk 4 above)
Comment (blacktear23) [reply Makefile:250]: Sure, I will remove it at last.

**Comment by tangenta** [comment]
Just thinking, is it possible to use [multi-schema change](https://github.com/pingcap/tidb/issues/14766) to implement this feature? Thus, we can introduce less changes:

```sql
create table t (a int, b int, index idx(a, b));
-- Query OK, 0 rows affected (0.17 sec)

alter table t drop index idx, drop column a, add index idx_1(b), rename index idx_1 to idx;
```

What we need to change is to lift the [restriction](https://github.com/pingcap/tidb/blob/master/docs/design/2022-04-15-multi-schema-change.md#mysql-compatibility:~:text=TiDB%20validates%20the%20schema%20changes%20against%20a%20snapshot%20schema%20structure%20retrieved%20before%20the%20execution%2C%20regardless%20of%20the%20previous%20changes%20in%20the%20same%20DDL%20statement.%20This%20may%20affect%20some%20common%20use%20cases.%20For%20example%2C%20ALTER%20TABLE%20t%20ADD%20COLUMN%20a%2C%20ADD%20INDEX%20i1(a)%3B%20is%20not%20supported.) for renaming index specifically.

**Comment by blacktear23** [comment]
For current implements I used to write it before, and the design is more friendly for cancelling and rollback. And drop column with composite index may cause create and remove more than 1 index at once. And I am not familiar with multi-schema change logic so I don't know whether it can handle thise logic.

**Pull request closed without merging**